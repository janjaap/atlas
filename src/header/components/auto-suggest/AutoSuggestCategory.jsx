import React from 'react';
import PropTypes from 'prop-types';
import AutoSuggestItem from './AutoSuggestItem';

const MAX_NUMBER_RESULTS = 3;

// Todo: Please consider rewriting the way we keep track on the active (selected) item in the
// autosuggest result list. Now we use an (arbitrary) high number for more results button ("..."),
// so this wont conflict with the activeSuggestion indexes.
export const MORE_RESULTS_INDEX = 999;

const AutoSuggestCategory = (props) => {
  const { category, activeSuggestion, query, onSuggestionSelection } = props;

  let suggestions = category.content;
  if (category.total_results > MAX_NUMBER_RESULTS) {
    suggestions = [...category.content, { label: '...', index: MORE_RESULTS_INDEX }];
  }

  return (
    <div className="auto-suggest__dropdown-category">
      <h4 className="auto-suggest__dropdown-category__heading qa-auto-suggest-header">
        {category.label}
      </h4>
      <ul>
        {suggestions.map((suggestion) => (
          <AutoSuggestItem
            key={suggestion.label + suggestion.index}
            isActive={activeSuggestion && activeSuggestion.index === suggestion.index}
            onSuggestionSelection={(e) => {
              onSuggestionSelection(suggestion, e);
            }}
            content={suggestion.label}
            query={query}
          />
        ))}
      </ul>
    </div>
  );
};

AutoSuggestCategory.defaultProps = {};

AutoSuggestCategory.propTypes = {
  activeSuggestion: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  category: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  onSuggestionSelection: PropTypes.func.isRequired,
  query: PropTypes.string.isRequired
};

export default AutoSuggestCategory;
