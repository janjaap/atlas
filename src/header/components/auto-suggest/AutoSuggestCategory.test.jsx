import React from 'react';
import { shallow } from 'enzyme';
import AutoSuggestCategory, { MORE_RESULTS_INDEX } from './AutoSuggestCategory';

describe('AutoSuggestCategory', () => {
  let props;
  beforeEach(() => {
    props = {
      activeSuggestion: {
        index: -1
      },
      category: {
        label: 'Straatnamen',
        content: [
          {
            category: 'Straatnamen',
            index: 0,
            label: 'Dam',
            uri: 'bag/openbareruimte/03630000003186/'
          },
          {
            category: 'Straatnamen',
            index: 1,
            label: 'Damloperspad',
            uri: 'bag/openbareruimte/03630000001038/'
          },
          {
            category: 'Straatnamen',
            index: 2,
            label: 'Damrak',
            uri: 'bag/openbareruimte/03630000003187/'
          }
        ],
        total_results: 6
      },
      onSuggestionSelection: jest.fn(),
      query: 'dam'
    };
  });

  it('should render ellipsis when there are more then 3 results', () => {
    const wrapper = shallow(
      <AutoSuggestCategory {...props} />
    );

    expect(wrapper).toMatchSnapshot();
    const items = wrapper.find('AutoSuggestItem');
    expect(items.length).toBe(4);
  });

  it('should handle the click on an ellipsis option', () => {
    const wrapper = shallow(
      <AutoSuggestCategory {...props} />
    );

    const mockEvent = { event: 'event' };
    const itemWrapper = wrapper.find('AutoSuggestItem').at(3).dive();
    itemWrapper.find('button').simulate('click', mockEvent);
    expect(props.onSuggestionSelection).toHaveBeenCalledWith({
      index: MORE_RESULTS_INDEX,
      label: '...'
    }, mockEvent);
  });

  it('should render no ellipsis when there are 3 or less results', () => {
    props.category.total_results = 3;
    const wrapper = shallow(
      <AutoSuggestCategory {...props} />
    );

    expect(wrapper).toMatchSnapshot();
    const items = wrapper.find('AutoSuggestItem');
    expect(items.length).toBe(3);
  });
});
