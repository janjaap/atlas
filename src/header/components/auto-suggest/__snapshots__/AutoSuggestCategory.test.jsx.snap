// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoSuggestCategory should render ellipsis when there are more then 3 results 1`] = `
<div
  className="auto-suggest__dropdown-category"
>
  <h4
    className="auto-suggest__dropdown-category__heading qa-auto-suggest-header"
  >
    Straatnamen
  </h4>
  <ul>
    <AutoSuggestItem
      content="Dam"
      isActive={false}
      key="Dam0"
      onSuggestionSelection={[Function]}
      query="dam"
    />
    <AutoSuggestItem
      content="Damloperspad"
      isActive={false}
      key="Damloperspad1"
      onSuggestionSelection={[Function]}
      query="dam"
    />
    <AutoSuggestItem
      content="Damrak"
      isActive={false}
      key="Damrak2"
      onSuggestionSelection={[Function]}
      query="dam"
    />
    <AutoSuggestItem
      content="..."
      isActive={false}
      key="...999"
      onSuggestionSelection={[Function]}
      query="dam"
    />
  </ul>
</div>
`;

exports[`AutoSuggestCategory should render no ellipsis when there are 3 or less results 1`] = `
<div
  className="auto-suggest__dropdown-category"
>
  <h4
    className="auto-suggest__dropdown-category__heading qa-auto-suggest-header"
  >
    Straatnamen
  </h4>
  <ul>
    <AutoSuggestItem
      content="Dam"
      isActive={false}
      key="Dam0"
      onSuggestionSelection={[Function]}
      query="dam"
    />
    <AutoSuggestItem
      content="Damloperspad"
      isActive={false}
      key="Damloperspad1"
      onSuggestionSelection={[Function]}
      query="dam"
    />
    <AutoSuggestItem
      content="Damrak"
      isActive={false}
      key="Damrak2"
      onSuggestionSelection={[Function]}
      query="dam"
    />
  </ul>
</div>
`;
